File: sdk/typescript/src/api/caesar.ts
import type { HttpClient } from "../client.js";
import type {
  CaesarBalance,
  CaesarGovernorParams,
  CaesarRewardInfo,
  CaesarRouteResult,
  CaesarTransactionList,
  CaesarWalletInfo,
} from "../types.js";

export class CaesarApi {
  constructor(private readonly http: HttpClient) {}

  async wallet(): Promise<CaesarWalletInfo> {
    return this.http.get<CaesarWalletInfo>("/api/v1/caesar/wallet");
  }

  async balance(): Promise<CaesarBalance> {
    return this.http.get<CaesarBalance>("/api/v1/caesar/balance");
  }

  async transactions(limit?: number): Promise<CaesarTransactionList> {
    const query = limit !== undefined ? `?limit=${limit}` : "";
    return this.http.get<CaesarTransactionList>(
      `/api/v1/caesar/transactions${query}`,
    );
  }

  async rewards(): Promise<CaesarRewardInfo> {
    return this.http.get<CaesarRewardInfo>("/api/v1/caesar/rewards");
  }

  async routePacket(
    destination: string,
    amountGrams: number,
  ): Promise<CaesarRouteResult> {
    return this.http.post<CaesarRouteResult>("/api/v1/caesar/route", {
      destination,
      amount_grams: amountGrams,
    });
  }

  async governorParams(): Promise<CaesarGovernorParams> {
    return this.http.get<CaesarGovernorParams>(
      "/api/v1/caesar/governor/params",
    );
  }
}

File: sdk/cpp/src/api/caesar.cpp
// Copyright 2026 Hypermesh Foundation. All rights reserved.
// Licensed under the Business Source License 1.1.

#include <hypermesh/api/caesar.hpp>

namespace hypermesh {

CaesarWalletInfo CaesarApi::wallet() const {
    auto j = http_.get("/api/v1/caesar/wallet");
    return j.get<CaesarWalletInfo>();
}

CaesarBalance CaesarApi::balance() const {
    auto j = http_.get("/api/v1/caesar/balance");
    return j.get<CaesarBalance>();
}

CaesarTransactionList CaesarApi::transactions(uint32_t limit) const {
    std::string path = "/api/v1/caesar/transactions";
    if (limit > 0) {
        path += "?limit=" + std::to_string(limit);
    }
    auto j = http_.get(path);
    return j.get<CaesarTransactionList>();
}

CaesarRewardInfo CaesarApi::rewards() const {
    auto j = http_.get("/api/v1/caesar/rewards");
    return j.get<CaesarRewardInfo>();
}

CaesarRouteResult CaesarApi::route_packet(const std::string& destination,
                                           double amount_grams) const {
    auto body = nlohmann::json{
        {"destination", destination},
        {"amount_grams", amount_grams}
    };
    auto j = http_.post("/api/v1/caesar/route", body);
    return j.get<CaesarRouteResult>();
}

CaesarGovernorParams CaesarApi::governor_params() const {
    auto j = http_.get("/api/v1/caesar/governor/params");
    return j.get<CaesarGovernorParams>();
}

} // namespace hypermesh
